### DISCUSSION TITLE
See pipeline yaml configuration in Web UI

### DISCUSSION THREAD

**Discussion by CiucurDaniel** [concourse/concourse#8241] [Ideas]
My idea/proposal is to have the following functionality:

When you click on a pipeline and have the view with the jobs with their inputs and outputs. Here I would also like to be able to have a button that I can click on and let's say on the right side a panel opens that has the current .yaml configuration that is applied.

---------------------------------------------
You can have the following scenario:

Developer X makes a chance to pipeline.yaml adds it to Git.
But developer X forgets to use `fly set-pipeline` in order to update the state on Concourse as well.

Now Git and Concourse have different state. And the best Developer Z can do is go to Fly 
do `fly login` then `fly set-pipeline` to make sure the changes were applied.

Instead, it would be nice if Developer Z could check this right from Concourse UI.
More simple and no need to switch from web to cli where you might as well need to change the target, switch git repo to the pipeline.yaml and so on.

-------------------------------------------

Therefore it would be nice to see right from concourse the actual applied configuration as yaml.

The jobs with inputs and outputs give you a nice overview for sure! 
But consider this:

| my-app-git |   ------>     |  my-job |

If the resource itself was changed yes I can clearly see it because it would turn into:

| my-other-app-git |   ------>     |  my-job |

But what if the change was only regarding the branch of my-app. I told a colleague to put `development` branch instead of `main`.

Further we can say that ok, you can click on the job, then on the task and check what triggered the pipeline. 
But then again this is also too much inspection you must do, and also won't be that easy when you have a `task` that runs a script with many commands let's say.

One problem I see with this is that we need to take care not to show secrets:

```yaml
-name: my-app
  type: git
  source:
    username: ((secret-username))
    password: ((secret-password))
```
Those should still be hidden even on the UI.

I will post a quickly made picture to illustrate the idea.

![proposal](https://user-images.githubusercontent.com/43583377/161954835-05f90c78-d624-483b-8c5a-28d80d8ac62b.png)

Thank you!

**Comment by taylorsilva** [comment]
Initial take; a read-only view of the current pipeline's config from the web UI seems like an okay idea. We don't want to encourage users to edit the yaml from within Concourse itself, but a read-only view feels fine for cases where someone wants to verify the current config.

The maintainers won't be picking up this idea but if someone in the community wants to take this on I'd say go for it!

Some other thoughts:
- Using `fly` it is possible to get the current pipeline yaml using `fly get-pipeline`. So the user could stay in their terminal and find the answer to the question your scenario proposes
- > But developer X forgets to use fly set-pipeline in order to update the state on Concourse as well.
  
  the current solution to this problem within Concourse is to not have users do `fly set-pipeline`. Use the [`set_pipeline` step](https://concourse-ci.org/set-pipeline-step.html) instead, that way git is always the source of truth and never "did human remember to do `fly set-pipeline`?"

**Reply by CiucurDaniel** [reply]
> a read-only view of the current pipeline's config from the web UI seems like an okay idea. We don't want to encourage users to edit the yaml from within Concourse itself, but a read-only view feels fine for cases where someone wants to verify the current config.

✅ Yes, forgot to mention this aspect, but yes definitely I want it to be `read-only`.

Well, hopefully there will be an engage from the community regarding this idea 🙏  .

I can try look into it myself as well 👀 . Do you have any recommendations for me to get started?
I'm new to contributing.

 Thank you for the answer! And also I'm keeping an eye open on this thread of discussion.

**Comment by lrstanley** [comment]
I've thought about this as well, as it would be a huge help to us, in additional to being able to intercept in the UI. Often times we have people come to us with assistance with their pipelines, and I have to edit-target and/or login, intercept, etc via the terminal, when I could just click a button to see what the config is and given pointers on the potential issues. This is also helpful for folks in our internal-community who want to help others, but not have it take a bunch of steps to do so.

Some additional thoughts:

1. What happens if someone hardcoded credentials into the YAML?
2. Should it be queried on page load, or on demand, in the UI?
   * If queried on page load, this means any hardcoded credentials (if used) will always be pulled due to the YAML being pulled, even if the user never uses the functionality within the UI.
3. Should it have similar functionality to `fly set-pipeline --check-creds`? I'm not sure what this would look like, or if it would even make sense, but if the goal is to help troubleshooting, I'd be interested in the idea of being able to tell if a value was pulled or not (not necessarily the value of it, but).
